src/compose/service.ts

```diff
@@ -702,24 +726,76 @@ export class Service {
 				log.debug('  Network changes detected');
 			}
 		}
-		return sameNetworks && sameConfig;
+
+		// Check the network mode separetely, as if it is a
+		// service: network mode, the container id needs to be
+		// checked against the running containers
+
+		// When this function is called, it's with the current
+		// state as a parameter and the target as the instance.
+		// We shouldn't rely on that because it's not enforced
+		// anywhere. For that reason we need to consider both
+		// network_modes in the correct way
+		let sameNetworkMode = false;
+		for (const [a, b] of [
+			[this.config.networkMode, service.config.networkMode],
+			[service.config.networkMode, this.config.networkMode],
```

Comment:
I keep thinking about this in bg. The comparison logic here is complex enough to deserve a separate `NetworkMode` class.

General idea:
```typescript

class NetworkMode {

  private constructor(
    public readonly type: 'host'|'bridge'|'none'|'service_ref'|'container_ref',
    public readonly serviceReference: string|undefined,
    public readonly containerReference: string|undefined
  )

  public static from(composeLiteral: string): NetworkMode {
    // Parse the literal and
    return new NetworkMode(type, serviceRef, containerRef);
  }

  public equals(another: NetworkMode): boolean {
    // Extremely easy to do now - just check if all 3 fields match.
    return this.type === another.type 
        && this.containerRef === another.containerRef 
        && this.serviceRef === another.containerRef;
  }

}

```

And then here we'd do
```typescript
  const sameNetworkMode = this.config.networkMode.equals(service.config.networkMode);
```

Reply:
The code you posted doesn't do the same thing that is necessary and this PR does. That being said the overall idea is a good one, but I'd like to handle it another time and get this out.